src/sage/algebras/letterplace/free_algebra_letterplace.pyx

```diff
@@ -591,20 +601,22 @@ cdef class FreeAlgebra_letterplace(Algebra):
         cdef int i, j, base, exp, var_ind
         cdef list out = []
         cdef list tmp
-        for i from 0<=i<nblocks:
+        cdef int N = len(self._names)
+        for i in range(nblocks):
             base = i*ngens
-            tmp = [(j, E[base+j]) for j in range(ngens) if E[base+j]]
+            tmp = [(j, E[base + j]) for j in range(ngens) if E[base + j]]
             if not tmp:
                 continue
             var_ind, exp = tmp[0]
             if len(tmp) > 1 or exp > 1:
                 raise NotImplementedError("\n  Apparently you tried to view the letterplace algebra with\n  shift-multiplication as the free algebra over a finitely\n  generated free abelian monoid.\n  In principle, this is correct, but it is not implemented, yet.")
-
+            if self._nb_slackvars and var_ind == N:
+                continue
```

Comment:
what is this new code doing? is there an example that covers it?

Reply:
There are two cases : with or without degrees argument. If there are degrees, there are "slack variables" and `_nb_slackvar` is set to 1. One slack variable appears at the end of each block in these loops. To properly get rid of them, we need this new code.